substrate/frame/revive/src/vm/evm.rs

```diff
@@ -91,20 +119,162 @@ pub fn call<'a, E: Ext>(bytecode: Bytecode, ext: &'a mut E, inputs: EVMInputs) -
 }
 
 /// Runs the EVM interpreter until it returns an action.
-fn run<WIRE: InterpreterTypes>(
-	interpreter: &mut Interpreter<WIRE>,
-	table: &revm::interpreter::InstructionTable<WIRE, DummyHost>,
+fn run<'a, E: Ext>(
+	interpreter: &mut Interpreter<EVMInterpreter<'a, E>>,
+	table: &revm::interpreter::InstructionTable<EVMInterpreter<'a, E>, DummyHost>,
 ) -> InterpreterResult {
 	let host = &mut DummyHost {};
 	loop {
 		let action = interpreter.run_plain(table, host);
 		match action {
-			InterpreterAction::Return(result) => return result,
-			InterpreterAction::NewFrame(_) => unimplemented!(),
+			InterpreterAction::Return(result) => {
+				log::info!("Return {:?}", result);
+				return result;
+			},
+			InterpreterAction::NewFrame(frame_input) => match frame_input {
+				FrameInput::Call(call_input) => run_call(interpreter, call_input),
+				FrameInput::Create(create_input) => run_create(interpreter, create_input),
+				FrameInput::Empty => unreachable!(),
+			},
 		}
 	}
 }
 
+fn run_call<'a, E: Ext>(
+	interpreter: &mut Interpreter<EVMInterpreter<'a, E>>,
+	call_input: Box<CallInputs>,
+) {
+	let meter = interpreter.extend.gas_meter_mut();
+
+	let callee: H160 = call_input.target_address.0 .0.into();
+
+	let precompile = <AllPrecompiles<E::T>>::get::<E>(&callee.as_fixed_bytes());
+	match &precompile {
+		Some(precompile) if precompile.has_contract_info() => {
+			if meter.charge(RuntimeCosts::PrecompileWithInfoBase).is_err() {
+				interpreter.halt(revm::interpreter::InstructionResult::OutOfGas);
+				return;
+			}
+		},
+		Some(_) =>
+			if meter.charge(RuntimeCosts::PrecompileBase).is_err() {
+				interpreter.halt(revm::interpreter::InstructionResult::OutOfGas);
+				return;
+			},
+		None => {
+			let costs = if call_input.scheme.is_delegate_call() {
+				RuntimeCosts::DelegateCallBase
+			} else {
+				RuntimeCosts::CallBase
+			};
+			if meter.charge(costs).is_err() {
+				interpreter.halt(revm::interpreter::InstructionResult::OutOfGas);
+				return;
+			}
+		},
+	};
+
+	let input = match &call_input.input {
+		CallInput::Bytes(bytes) => bytes.to_vec(),
+		// Consider the usage fo SharedMemory as REVM is doing
+		CallInput::SharedBuffer(range) => interpreter.memory.global_slice(range.clone()).to_vec(),
+	};
```

Comment:
yeah I guess we could refactor call to use &[u8] instead of passing a Vec, I think for now we only ever construct the first variant anyway

Reply:
It looks more like an optimization. Let's fix this later.

Reply:
yeah and from what I read in the original PR it did not yield great optimisations